proposals/massive_context_refactor.cpp

```diff
@@ -0,0 +1,381 @@
+                return NULL;
+
+            void* p = _allocator.allocate(size);
+            if (p == NULL)
+                _allocator.rebalance(); 
+        }
+        static void  deallocate(void* p) { ... }
+        static void  bind(void* ptr, utensor::Tensor* hndl) {
+            _allocator.bind(ptr, hndl);
+        }
+
+};
+
+using utensor::DefaultTensorMetaDataAllocator = utensor::ArenaAllocator<512>;
+using utensor::DefaultRamTensorAllocator = utensor::ArenaAllocator<4096>;
+
+// Tensors also appear on the same heap as the Tensor metadata. This way we can move tensors around and delete them without affecting user code
+//template <typename Allocator=utensor::DefaultTensorMetaDataAllocator>
+class Tensor {
+    private:
+        utensor::TensorInterface* _ptr;
+        Tensor(const Tensor& that) {} // Cannot copy Tensors, must pass by reference
+
+    public:  
+        utensor::TensorInterface* operator->(0) { return _ptr; }
+        Tensor(utensor::TensorInterface* ptr) : _ptr(ptr) {
+            Context::DefaultTensorMetaDataAllocator::bind(this, ptr);
+        }
+        // Add some bits to make the interface nicer to the user
+
+        // Force everything to be on the utensor allocator
+        void* operator new(size_t sz) { // Have to delegate this size from tensors somehow + sizeof(Tensor)
+            void* p = Context::DefaultTensorMetaDataAllocator::allocate(sz); 
+            return p;
+        }
+        void operator delete(void* p) {
+            Context::DefaultTensorMetaDataAllocator::deallocate(p);
+        }
+
+        // KEY BIT
+        friend class utensor::AllocatorInterface;
+};
+
+// sizeof(NamedTensorReference) == 8 bytes
+class NamedTensorReference {
+    public:
+    const utensor::string& name; //Fixed
+    utensor::Tensor& tensor;     //Modifiable
+    
+    NamedTensorReference(const utensor::string& name, utensor::Tensor& tensor) : name(name), tensor(tensor) {
+        Context& ctx = Context::get_default_context();
+        ctx.push(*this);
+    }
+    
+};
+
+// Same as Named Tensor but not registered in the context class 
+class SimpleNamedTensor {
+    public:
+    const utensor::string& name; //Fixed
+    utensor::Tensor& tensor;     //Modifiable
+    
+    SimpleNamedTensor(const utensor::string& name, utensor::Tensor& tensor) : name(name), tensor(tensor) {}
+};
+
+// Tensor maps are fixed size to force input output mismatched errors
+class TensorMapInterface {
+public:
+    virtual SimpleNamedTensor& operator[](const utensor::string& name) = 0;
+    virtual const SimpleNamedTensor& operator[](const utensor::string& name) const = 0;
+    static SimpleNamedTensor not_found(utensor::string("NotFound"), static_cast<utensor::Tensor>(NULL));
+};
+
+template<size_t size>
+class FixedTensorMap : public TensorMapInterface{
+public:
+    TensorMap(SimpleNamedTensor map[size]) : _map[map] {}
+    virtual ~TensorMap() {}
+    SimpleNamedTensor& operator[](const utensor::string& name){
```

Comment:
We can do better than a linear search?   If this class serves as a dict, we should be able to map `name` to tensor* or tensor-handle itself. We then may not need the `simple named tensor` container.

Reply:
I see.  It looks like every op has its own tensor maps. Given the die of the maps is small, linear search is ok.
Similar functionality may be implemented by declaring inner structures within an op, as proposed by the original issue. Perhaps there is a hybrid approach to this.